base/iterators.jl

```diff
@@ -958,4 +959,105 @@ function next(itr::PartitionIterator, state)
     return resize!(v, i), state
 end
 
+"""
+    Stateful(itr)
+
+There are several different ways to think about this iterator wrapper:
+    1. It provides a mutable wrapper around an iterator and
+       its iteration state.
+    2. It turns an iterator-like abstraction into a Channel-like
+       abstraction.
+    3. It's an iterator that mutates to become its own rest iterator
+       whenever an item is produced.
+
+`Stateful` provides the regular iterator interface. Like other mutable iterators
+(e.g. `Channel`), if iteration is stopped early (e.g. by a `break` in a `for` loop),
+iteration can be resumed from the same spot by continuing to iterate over the
+same iterator object (in contrast, an immutable iterator would restart from the
+beginning).
+
+# Example:
+```jldoctest
+julia> a = Iterators.Stateful("abcdef");
+
+julia> isempty(a)
+false
+
+julia> popfirst!(a)
+'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)
+
+julia> collect(Iterators.take(a, 3))
+3-element Array{Char,1}:
+ 'b'
+ 'c'
+ 'd'
+
+julia> collect(a)
+2-element Array{Char,1}:
+ 'e'
+ 'f'
+```
+
+```jldoctest
+julia> a = Iterators.Stateful([1,1,1,2,3,4]);
+
+julia> for x in a; x == 1 || break; end
+
+julia> Base.peek(a)
+3
+
+# Sum the remaining elements
+julia> sum(a)
+7
+````
+"""
+mutable struct Stateful{T, VS}
+    itr::T
+    # A bit awkward right now, but adapted to the new iteration protocol
+    nextvalstate::Union{VS, Nothing}
+    taken::Int
+    # Try to find an appropriate type for the (value, state tuple).
+    # To account for iterators with initial state of different type,
+    # we refine twice. However, if that doesn't fixpoint, we just jump
+    # to `Any`. In normal circumstances, the compiler should be eliminating
+    # this entire allocation anyway so the type doesn't matter. This is just
+    # an optimization to avoid a performance cliff in cases where the compiler
+    # does not.
+    function approx_iter_type(itrT::Type, VS::Type)
+        VS = typejoin(VS, Base._return_type(next, Tuple{itrT,
+            tuple_type_head(tuple_type_tail(VS))}))
+        VS = typejoin(VS, Base._return_type(next, Tuple{itrT,
+            tuple_type_head(tuple_type_tail(VS))}))
+        !(VS <: Tuple{Any, Any}) && return Tuple{Any, Any}
+        Base._return_type(next, Tuple{itrT, tuple_type_head(tuple_type_tail(VS))}) <: VS ? VS : Tuple{Any, Any}
```

Comment:
Actually, maybe define this recursively, so that you're not having to deal with copying the code and conditional tests (like the several missing is-subtype NTuple{2, Any} tests above). Also, calling this after calling `next(start)` also means you're giving it type-unstable input, which I would expect is performing very poorly. It'll be better to call this on the result of `start(itr)`.
```julia
function approx_iter_type(itrT::Type, state::Type)
    nextstate = Base._return_type(next, Tuple{itrT, state})
    nextstate <: Tuple{Any, Any} || return Any
    nextstate = typejoin(state, fieldtype(nextstate, 2))
    return (state == nextstate ? state : approx_iter_type(itrT, nextstate))
end
```

Also, `Tuple{Any, Any}` sounds nice, and is very precise, but can also perform much worse at runtime than just returning `Any` (since it'll force us to emit extra code to try to check + convert everything).

I'm not sure how we'll make this work after switching to the new, type-unstable iteration protocol. I think we can get inference to union-split the inference call to `fieldtype`, but it may be tricky to `typejoin` the result – perhaps we should be only unrolling the loop only exactly once to avoid that call:

```julia
approx_iter_type(itrT::Type) = approx_iter_type(itr, Base._return_type(iterate, Tuple{itrT}))
function approx_iter_type(itrT::Type, vstate::Type)
    vstate <: Union{Nothing, Tuple{Any, Any}} || return Any
    nextvstate = Base._return_type(next, Tuple{itrT, fieldtype(vstate, 2)})
    return (vstate <: nextvstate ? nextvstate : (nextvstate <: vstate ? vstate : Any))
end
```
Since anyways, if we don't pretty much get a concrete type right away, there's likely very little to be gained from continuing to try. (considering also the structure of a for loop, which also only splits off the first element, and the storage, which will only be efficient if it always has a nearly-concrete type)